src/backend/wayland/application.rs

```diff
@@ -14,582 +14,222 @@
 pub struct Application {
-    pub(super) data: std::sync::Arc<Data>,
-}
-
-#[allow(dead_code)]
-pub(crate) struct Data {
-    pub(super) wayland: std::rc::Rc<display::Environment>,
-    pub(super) zwlr_layershell_v1: Option<wl::Main<ZwlrLayerShellV1>>,
-    pub(super) wl_compositor: wl::Main<WlCompositor>,
-    pub(super) wl_shm: wl::Main<WlShm>,
-    /// A map of wayland object IDs to outputs.
-    ///
-    /// Wayland will update this if the output change. Keep a record of the `Instant` you last
-    /// observed a change, and use `Output::changed` to see if there are any newer changes.
-    ///
-    /// It's a BTreeMap so the ordering is consistent when enumerating outputs (not sure if this is
-    /// necessary, but it negligible cost).
-    pub(super) outputs: Rc<RefCell<BTreeMap<u32, outputs::Meta>>>,
-    pub(super) seats: Rc<RefCell<BTreeMap<u32, Rc<RefCell<Seat>>>>>,
-
-    /// Handles to any surfaces that have been created.
-    pub(super) handles: RefCell<im::OrdMap<u64, WindowHandle>>,
-
-    /// Available pixel formats
-    pub(super) formats: RefCell<Vec<wl_shm::Format>>,
-    /// Close flag
-    pub(super) shutdown: Cell<bool>,
-    /// The currently active surface, if any (by wayland object ID)
-    pub(super) active_surface_id: RefCell<std::collections::VecDeque<u64>>,
-    // Stuff for timers
-    /// A calloop event source for timers. We always set it to fire at the next set timer, if any.
-    pub(super) timer_handle: calloop::timer::TimerHandle<TimerToken>,
-    /// We stuff this here until the event loop, then `take` it and use it.
-    timer_source: RefCell<Option<calloop::timer::Timer<TimerToken>>>,
-    /// Currently pending timers
-    ///
-    /// The extra data is the surface this timer is for.
-    pub(super) timers: RefCell<BinaryHeap<Timer>>,
-
-    pub(super) roundtrip_requested: RefCell<bool>,
-
-    /// track if the display was flushed during the event loop.
-    /// prevents double flushing unnecessarily.
-    pub(super) display_flushed: RefCell<bool>,
-    /// reference to the pointer events manager.
-    pub(super) pointer: pointers::Pointer,
-    /// reference to the keyboard events manager.
-    keyboard: keyboard::Manager,
-    clipboard: clipboard::Manager,
-    // wakeup events when outputs are added/removed.
-    outputsqueue: RefCell<Option<calloop::channel::Channel<outputs::Event>>>,
+    // `State` is the items stored between `new` and `run`
+    // It is stored in an Rc<RefCell>> because Application must be Clone
+    // The inner is taken in `run`
+    state: Rc<RefCell<Option<WaylandState>>>,
+    pub(super) compositor: wl_compositor::WlCompositor,
+    pub(super) wayland_queue: QueueHandle<WaylandState>,
+    pub(super) xdg_shell: Weak<XdgShell>,
+    loop_handle: LoopHandle<'static, WaylandState>,
+    loop_signal: LoopSignal,
+    pub(super) idle_sender: Sender<IdleAction>,
+    pub(super) loop_sender: channel::Sender<ActiveAction>,
+    pub(super) raw_display_handle: *mut c_void,
```

Comment:
You could keep the `Connection` around for this or get the underlying `Backend` from wayland-backend and keep that.

Reply:
I would like to avoid random 'Application' instances keeping the connection alive(?) without reason, which it's possible that this would do.

But I'm not tied to this particular implementation decision